pkg/kubelet/cm/topologymanager/scope.go

```diff
@@ -125,6 +125,9 @@ func (s *scope) admitPolicyNone(pod *v1.Pod) lifecycle.PodAdmitResult {
 	for _, container := range append(pod.Spec.InitContainers, pod.Spec.Containers...) {
 		err := s.allocateAlignedResources(pod, &container)
 		if err != nil {
+			if _, ok := err.(*SMTAlignmentError); ok {
+				return smtAlignmentError(err)
+			}
```

Comment:
It's unfortunate that this needs to be type checked like this in so many places (I count at least 4). Is there some better way to avoid this?

Reply:
This is potentially related to https://github.com/kubernetes/kubernetes/pull/101432#discussion_r623354087 .
The reason why this block is replicated is because there indeed are 4 places on which we do the admission control:
1. container manager (topology manager disabled)
2. topology manager, when enabled, delegated by container manager; if the policy is none, then topology manager itself does the admission controls otherwise it further delegates to the 
3. scope_container in topology manager or
4. scope_pod in topology manager

Additionally, `SMTAlignmentError` bleongs in cpumanager, but the specific resource managers need to import the `topologymanager` package because of the `HintProvider` interface.

A way to improve both sides could be to run the actual admission check in the resource manager, but to build the actual `lifecycle.PodAdmitResult` always in a single place, and the most suitable seems to be the container manager.
IOW, the container manager will keep delegating the actual checks - which the specific resource managers know how to do - but will centralize the admission control proper.

This way we deduplicate all the `lifecycle.PodAdmitResult` creation bits and we have a cleaner import hierarchy - container manager needs to import both `topologymanager` and `cpumanager` already!
However unfortunately this change will require some refactoring to fix the managers interface. I prototyped this change here: https://gist.github.com/fromanirh/33d2094a8fdf23fdcaf9e38f7003bb28

Could be this a good direction forward?